python/dgl/data/explain_syn.py

```diff
@@ -0,0 +1,757 @@
+    - Attach the motifs to randomly selected nodes of the base graph.
+    - Perturb the graph by adding random edges.
+    - Generate constant feature for all nodes, which is 1.
+    - Nodes in the tree belong to class 0 and nodes in cycles belong to class 1.
+
+    Parameters
+    ----------
+    tree_height : int, optional
+        Height of the balanced binary tree. Default: 8
+    num_motifs : int, optional
+        Number of cycle motifs to use. Default: 60
+    perturb_ratio : float, optional
+        Number of random edges to add in perturbation divided by the
+        number of original edges in the graph. Default: 0.01
+    seed : integer, random_state, or None, optional
+        Indicator of random number generation state. Default: None
+    raw_dir : str, optional
+        Raw file directory to store the processed data. Default: ~/.dgl/
+    force_reload : bool, optional
+        Whether to always generate the data from scratch rather than load a cached version.
+        Default: False
+    verbose : bool, optional
+        Whether to print progress information. Default: True
+    transform : callable, optional
+        A transform that takes in a :class:`~dgl.DGLGraph` object and returns
+        a transformed version. The :class:`~dgl.DGLGraph` object will be
+        transformed before every access. Default: None
+
+    Attributes
+    ----------
+    num_classes : int
+        Number of node classes
+
+    Examples
+    --------
+
+    >>> from dgl.data import TreeCycles
+    >>> dataset = TreeCycles()
+    >>> dataset.num_classes
+    2
+    >>> g = dataset[0]
+    >>> label = g.ndata['label']
+    >>> feat = g.ndata['feat']
+    """
+    def __init__(self,
+                 tree_height=8,
+                 num_motifs=60,
+                 perturb_ratio=0.01,
+                 seed=None,
+                 raw_dir=None,
+                 force_reload=False,
+                 verbose=True,
+                 transform=None):
+        self.tree_height = tree_height
+        self.num_motifs = num_motifs
+        self.perturb_ratio = perturb_ratio
+        self.seed = seed
+        super(TreeCycles, self).__init__(name='TREE-CYCLES',
+                                         url=None,
+                                         raw_dir=raw_dir,
+                                         force_reload=force_reload,
+                                         verbose=verbose,
+                                         transform=transform)
+
+    def process(self):
+        if self.seed is not None:
+            np.random.seed(self.seed)
+
+        g = nx.balanced_tree(r=2, h=self.tree_height)
+        edges = list(g.edges())
+        src, dst = map(list, zip(*edges))
+        n = nx.number_of_nodes(g)
+
+        # Nodes in the base tree graph belong to class 0
+        node_labels = [0] * n
+        # The motifs will be evenly attached to the nodes in the base graph.
+        spacing = math.floor(n / self.num_motifs)
+
+        cycle_size = 6
```

Comment:
Consider exposing this parameter

Reply:
This is the hyperparameter for the motif. The question is then if you also want to expose the motif hyperparameter for the other datasets.

Reply:
Done.